### PR TITLE
for pull request 1

### PR DESCRIPTION
**Pull request by AndreiKeino**
pull request от Андрей Кейно

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,4 +3,6 @@ dist/
 *.egg-info/
 __pycache__
 .pytest_cache
-setup.bat
\ No newline at end of file
+setup.bat
+.idea
+.vscode
```
Comment (blackyblack) [inline .gitignore:8]: Ignored files should not present in repository. Either remove .vscode before ignoring it, or remove this change.

File: symplyphysics/laws/relativistic/moving_particle_energy.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from sympy import sqrt
+from sympy import Eq, solve
+from sympy.physics.units import speed_of_light as c, convert_to
```
Comment (blackyblack) [inline symplyphysics/laws/relativistic/moving_particle_energy.py:3]: Do not rename speed of light.

File: symplyphysics/laws/relativistic/moving_particle_energy.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+from symplyphysics import units, Quantity, Symbol, print_expression, validate_input, validate_output
+
+# Description
+## The relativistic energy of a particle of rest mass m moving in 
+# your frame of reference at speed v 
+
+# Law: E = gamma * m * c**2, where
+## E is energy of the particle
+## m is the rest mass of the particle
+## c is the speed of light.
```
Comment (blackyblack) [inline symplyphysics/laws/relativistic/moving_particle_energy.py:13]:
```suggestion
## c is the speed of light
```

File: symplyphysics/laws/relativistic/moving_particle_energy.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+def print_law() -> str:
+    return print_expression(law)
+
+
+@validate_input(rest_mass_=rest_mass, particle_veclocity_=particle_velocity)
+@validate_output(particle_energy)
+
+def calculate_particle_energy(rest_mass_: Quantity,
+                              particle_velocity_: Quantity) -> Quantity:
+    if (convert_to(particle_velocity_, [units.meter, units.second]) >=
```
Comment (blackyblack) [inline symplyphysics/laws/relativistic/moving_particle_energy.py:32]: Remove this check.

File: test/relativistic/moving_particle_energy_test.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import pytest
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:1]: See existing tests for reference.

File: test/relativistic/moving_particle_energy_test.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+from symplyphysics.laws.relativistic.moving_particle_energy import calculate_particle_energy
+from sympy.physics.units import (speed_of_light, 
+                                 meter, 
+                                 second, 
+                                 kg, 
+                                 convert_to)
+import math
+
+def test_basic():
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:10]: See other tests for test naming and structure.

File: test/relativistic/moving_particle_energy_test.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    e = calculate_particle_energy(1 * kg, 0.5 * speed_of_light)
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:11]: You should use Quantity for parameters.

File: test/relativistic/moving_particle_energy_test.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    check = 1 * kg * convert_to(
+        speed_of_light, [meter, second]) ** 2 * \
+        math.sqrt(1 / (1 - 0.5 ** 2))
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:14]: Use sympy sqrt.

File: test/relativistic/moving_particle_energy_test.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    assert (convert_to(e, [meter, kg, second])) == check
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:15]: Use approximate comparison.

File: test/relativistic/moving_particle_energy_test.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        res = calculate_particle_energy(1 * kg, 1.0 * speed_of_light)
+    with pytest.raises(ValueError) as e:
+        res = calculate_particle_energy(1 * kg, 1.1 * speed_of_light)
+
+
+# run from ... simplyphysics/symplyphysics dir
+# python -m pytest test/relativistic/moving_particle_energy_test.py        
+
+        
+if __name__ == "__main__":
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:29]: Remove this and comment above.
Run tests with pytest.

**Comment by AndreiKeino** [comment]
Are you sure about this?

In symplyphysics/laws/relativistic/moving_particle_energy.py
<https://github.com/blackyblack/symplyphysics/pull/134#discussion_r1436965737>
:

> +particle_energy = Symbol("particle_energy", units.energy)
+rest_mass = Symbol("rest_mass", units.mass)
+particle_velocity = Symbol("particle_velocity", units.velocity)
+law = Eq(particle_energy, 1 / sqrt(1 - (particle_velocity / c) ** 2)
* rest_mass * c ** 2)
+
+
+def print_law() -> str:
+    return print_expression(law)
+
+
***@***.***_input(rest_mass_=rest_mass, particle_veclocity_=particle_velocity)
***@***.***_output(particle_energy)
+
+def calculate_particle_energy(rest_mass_: Quantity,
+                              particle_velocity_: Quantity) -> Quantity:
+    if (convert_to(particle_velocity_, [units.meter, units.second]) >=

Remove this check.

**Comment by blackyblack** [comment]
> Are you sure about this? In symplyphysics/laws/relativistic/moving_particle_energy.py <[#134 (comment)](https://github.com/blackyblack/symplyphysics/pull/134#discussion_r1436965737)> :
> +particle_energy = Symbol("particle_energy", units.energy)
> +rest_mass = Symbol("rest_mass", units.mass) +particle_velocity = Symbol("particle_velocity", units.velocity) +law = Eq(particle_energy, 1 / sqrt(1 - (particle_velocity / c) ** 2) * rest_mass * c ** 2) + + +def print_law() -> str: + return print_expression(law) + + ***@***.***_input(rest_mass_=rest_mass, particle_veclocity_=particle_velocity) ***@***.***_output(particle_energy) + +def calculate_particle_energy(rest_mass_: Quantity, + particle_velocity_: Quantity) -> Quantity: + if (convert_to(particle_velocity_, [units.meter, units.second]) >= Remove this check.

Yes. Faster than light values are not directly prohibited by the law and user will see his mistake from the output of the function.

**Comment by AndreiKeino** [comment]
I've made a next pull request. How long will the test period last?

File: symplyphysics/laws/relativistic/moving_particle_energy.py [hunk 11]
```diff
@@ -1,23 +1,23 @@
...
-from sympy.physics.units import speed_of_light as c, convert_to
+from sympy.physics.units import speed_of_light, convert_to
 from symplyphysics import units, Quantity, Symbol, print_expression, validate_input, validate_output
 
 # Description
 ## The relativistic energy of a particle of rest mass m moving in 
 # your frame of reference at speed v 
 
-# Law: E = gamma * m * c**2, where
+# Law: E = gamma * m * speed_of_light**2, where
```
Comment (blackyblack) [inline symplyphysics/laws/relativistic/moving_particle_energy.py:10]: You could leave `c` here.

File: test/relativistic/moving_particle_energy_test.py [hunk 12]
```diff
@@ -1,31 +1,33 @@
...
+    convert_to,
+)
+from sympy import Abs, sqrt
+from sympy.physics.units import speed_of_light
 
 
-# run from ... simplyphysics/symplyphysics dir
-# python -m pytest test/relativistic/moving_particle_energy_test.py        
+def test_basic():
+    e = calculate_particle_energy(Quantity(1 * units.kilogram),
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:17]: Put quantities in test fixture, like in other tests.

File: test/relativistic/moving_particle_energy_test.py [hunk 13]
```diff
@@ -1,31 +1,33 @@
...
 
 
-# run from ... simplyphysics/symplyphysics dir
-# python -m pytest test/relativistic/moving_particle_energy_test.py        
+def test_basic():
+    e = calculate_particle_energy(Quantity(1 * units.kilogram), 
+                                  Quantity(0.5 * speed_of_light))
+    assert SI.get_dimension_system().equivalent_dims(e.dimension, units.energy)
+    e_nondim = convert_to(e, 
+                          units.kilogram * units.meter ** 2 / units.second ** 2).as_coeff_Mul()[0]
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:21]: Why not joules?

File: test/relativistic/moving_particle_energy_test.py [hunk 14]
```diff
@@ -1,31 +1,33 @@
...
+    c = convert_to(speed_of_light, units.meter / units.second).as_coeff_Mul()[0]
+    check = c ** 2 * sqrt(1 / (1 - 0.5 ** 2))
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:23]: You are basically copypasting the formula from the law. What if it was incorrect? You won't detect an error with this test.
Look on the internet for some real value to put here.

File: test/relativistic/moving_particle_energy_test.py [hunk 15]
```diff
@@ -1,31 +1,33 @@
...
+    assert Abs(e_nondim - check) < 1e-16
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:24]: You've imported `approx` but do not use it. Look in other tests how to use it.

File: test/relativistic/moving_particle_energy_test.py [hunk 16]
```diff
@@ -1,31 +1,33 @@
...
+    
 
-        
-if __name__ == "__main__":
-    test_basic()
-    test_exception()
\ No newline at end of file
+def test_bad_mass():
```
Comment (blackyblack) [inline test/relativistic/moving_particle_energy_test.py:27]: What about bad velocity test?

File: .gitignore [hunk 17]
```diff
@@ -4,5 +4,4 @@ dist/
 __pycache__
 .pytest_cache
 setup.bat
-.idea
-.vscode
```
Comment (blackyblack) [inline .gitignore]: Now you've deleted .vscode from repo, put it back to .gitignore

**Comment by AndreiKeino** [comment]
*What about bad velocity test? *
It makes sense.

*You've imported approx but do not use it. Look in other tests how to use
it*
.It doesn't make sense to use approx. If there is an error, approx gives an
uninformative message.

*You are basically copypasting the formula from the law. What if it was
incorrect? You won't detect an error with this test. Look on the internet
for some real value to put here*
it doesn't make sense at all. I am quite able to calculate a real value.

And so on...

Most of these comments don't make sense. I don't interested in continuing.

**Pull request closed without merging**